keepalived-vip/README.md

```diff
@@ -0,0 +1,374 @@
+
+## Disclaimer:
+- This is a **work in progress**.
+
+## Overview
+
+There are 2 ways to expose a service in the current kubernetes service model:
+
+- Create a cloud load balancer.
+- Allocate a port (the same port) on every node in your cluster and proxy traffic through that port to the endpoints.
+
+This just works. What's the issue then? 
+
+The issue is that it does not provide High Availability because beforehand is required to know the IP addresss of the node where is running and in case of a failure the pod can be be moved to a different node. Here is where ipvs could help. 
+The idea is to define an IP address per service to expose it outside the Kubernetes cluster and use vrrp to announce this "mapping" in the local network.
+With 2 or more instance of the pod running in the cluster is possible to provide high availabity using a single IP address.
+
+##### What is the difference between this and [service-loadbalancer](https://github.com/kubernetes/contrib/tree/master/service-loadbalancer) or [nginx-alpha](https://github.com/kubernetes/contrib/tree/master/Ingress/controllers/nginx-alpha) to expose one or more services?
+
+This should be considered a complement, not a replacement for HAProxy or nginx. The goal using keepalived is to provide high availability and to bring certainty about how an exposed service can be reached. For instance keepalived can use used to expose the service-loadbalancer or nginx ingress controller in the LAN using one IP address.
+
+
+
+## Requirements
+
+[Daemonsets](https://github.com/kubernetes/kubernetes/blob/master/docs/design/daemon.md) enabled is the only requirement. Check this [guide](https://github.com/kubernetes/kubernetes/blob/master/docs/api.md#enabling-resources-in-the-extensions-group) with the required flags in kube-apiserver.
+
+
+
+## Configuration
+
+To expose a service add the annotation `k8s.io/public-vip` in the service with the IP address to be use. This IP must be routable inside the LAN and must be available.
+By default the IP address of the pods are used to route the traffic. This means that is one pod dies or a new one is created by a scale event the keepalived configuration file will be updated and reloaded.
+
+
+
+## Example
+
+First we create a new replication controller and service
+```
+$ kubectl create -f examples/echoheaders.yaml
+replicationcontroller "echoheaders" created
+You have exposed your service on an external port on all nodes in your
+cluster.  If you want to expose this service to the external internet, you may
+need to set up firewall rules for the service port(s) (tcp:30302) to serve traffic.
+
+See http://releases.k8s.io/HEAD/docs/user-guide/services-firewalls.md for more details.
+service "echoheaders" created
+```
+
+Next add the required annotation to expose the service using a local IP
+
+```
+$ kubectl annotate svc echoheaders "k8s.io/public-vip=10.4.0.50"
```

Comment:
It would be really cool if you could give an ingress an public vip like this as well. The Ingress was born out of the need to fanout a single ip to various services, either based on hostname or url regex.